File: create_asa.py
#!/usr/bin/env python3
"""
Algorand ASA Creation Script
Creates an ASA on TestNet using a 24-word BIP39 mnemonic
"""

import algosdk
from algosdk.v2client import algod
from algosdk.transaction import AssetCreateTxn, wait_for_confirmation
from bip_utils import Bip39SeedGenerator, Bip44, Bip44Coins, Bip44Changes
import base64

def derive_algorand_keypair_from_bip39(mnemonic_words):
    """
    Derive Algorand keypair from 24-word BIP39 mnemonic
    Uses BIP44 derivation path: m/44'/283'/0'/0/0 (283 is Algorand's coin type)
    """
    try:
        # Generate seed from mnemonic
        seed_bytes = Bip39SeedGenerator(mnemonic_words).Generate()
        
        # Create BIP44 context for Algorand (coin type 283)
        bip44_mst_ctx = Bip44.FromSeed(seed_bytes, Bip44Coins.ALGORAND)
        
        # Derive account 0, change 0, address 0
        bip44_acc_ctx = bip44_mst_ctx.Purpose().Coin().Account(0)
        bip44_chg_ctx = bip44_acc_ctx.Change(Bip44Changes.CHAIN_EXT)
        bip44_addr_ctx = bip44_chg_ctx.AddressIndex(0)
        
        # Get the private key bytes (32 bytes for Ed25519)
        private_key_bytes = bip44_addr_ctx.PrivateKey().Raw().ToBytes()
        
        # Create Algorand account from private key
        account = algosdk.account.Account(private_key_bytes)
        
        return account.private_key, account.address
        
    except Exception as e:
        print(f"❌ Error deriving keypair: {e}")
        raise

def create_asa_on_testnet(private_key, address):
    """
    Create an ASA on Algorand TestNet
    """
    # TestNet Algod client (AlgoExplorer API)
    algod_address = "https://testnet-api.algonode.cloud"
    algod_token = ""
    
    # Initialize the algod client
    algod_client = algod.AlgodClient(algod_token, algod_address)
    
    try:
        # Get network suggested parameters
        params = algod_client.suggested_params()
        
        # Asset creation parameters
        asset_create_txn = AssetCreateTxn(
            sender=address,
            sp=params,
            total=1000,           # Total supply
            default_frozen=False, # Assets are not frozen by default
            unit_name="TIX",      # Unit name
            asset_name="AUTHENTIX_EVENT", # Asset name
            manager=address,      # Manager address
            reserve=address,      # Reserve address
            freeze=address,       # Freeze address
            clawback=address,     # Clawback address
            decimals=0            # Number of decimals (0 for NFT-like tickets)
        )
        
        # Sign the transaction
        signed_txn = asset_create_txn.sign(private_key)
        
        # Submit the transaction
        txid = algod_client.send_transaction(signed_txn)
        print(f"⏳ Transaction sent with ID: {txid}")
        
        # Wait for confirmation
        confirmed_txn = wait_for_confirmation(algod_client, txid, 4)
        
        # Get the asset ID from the transaction
        asset_id = confirmed_txn["asset-index"]
        
        return asset_id, txid
        
    except Exception as e:
        print(f"❌ Error creating ASA: {e}")
        raise

def main():
    """
    Main function to create ASA from BIP39 mnemonic
    """
    # Your 24-word BIP39 mnemonic
    mnemonic = "crowd awkward correct unable tube analyst wash mind collect crumble trip legal increase medal segment vapor stay battle scene retreat later destroy battle vacuum"
    
    print("🚀 AuthenTIX ASA Creation Script")
    print("=" * 50)
    
    # Step 1: Derive Algorand keypair from BIP39 mnemonic
    print("📝 Deriving Algorand keypair from BIP39 mnemonic...")
    try:
        private_key, address = derive_algorand_keypair_from_bip39(mnemonic)
        print(f"✅ Derived Address: {address}")
    except Exception as e:
        print(f"❌ Failed to derive keypair: {e}")
        return
    
    # Step 2: Check account balance
    print("\n💰 Checking account balance...")
    try:
        algod_client = algod.AlgodClient("", "https://testnet-api.algonode.cloud")
        account_info = algod_client.account_info(address)
        balance = account_info.get('amount', 0) / 1_000_000  # Convert microAlgos to Algos
        print(f"✅ Account Balance: {balance} ALGO")
        
        if balance < 0.1:
            print("⚠️  Warning: Low balance. You need at least 0.1 ALGO for transaction fees.")
            print("💡 Get TestNet funds from: https://testnet.algoexplorer.io/dispenser")
    except Exception as e:
        print(f"⚠️  Could not check balance: {e}")
    
    # Step 3: Create ASA
    print("\n🎫 Creating ASA on TestNet...")
    try:
        asset_id, txid = create_asa_on_testnet(private_key, address)
        
        print("\n" + "=" * 50)
        print("🎉 ASA CREATED SUCCESSFULLY!")
        print("=" * 50)
        print(f"🆔 Asset ID: {asset_id}")
        print(f"📋 Transaction ID: {txid}")
        print(f"🔗 View on AlgoExplorer: https://testnet.algoexplorer.io/asset/{asset_id}")
        print(f"🔗 Transaction: https://testnet.algoexplorer.io/tx/{txid}")
        print("=" * 50)
        
        # Save asset ID to file for easy reference
        with open("asset_id.txt", "w") as f:
            f.write(str(asset_id))
        print(f"💾 Asset ID saved to: asset_id.txt")
        
    except Exception as e:
        print(f"❌ Failed to create ASA: {e}")
        return

if __name__ == "__main__":
    main()

File: scripts/create-asa.js
const algosdk = require('algosdk');

// Configuration from your .env.local
const ALGOD_URL = "https://testnet-api.algonode.cloud";
const ORGANIZER_WALLET_ADDRESS = "OG7MBB2QNNORL3F3RB3IMBECKLXE3IZRKXCQGDAXOVHGDGFVR2PC6BCTBI";

// Initialize Algorand client
const algodClient = new algosdk.Algodv2('', ALGOD_URL, '');

async function createASA() {
  try {
    console.log('🎫 Creating AuthenTIX Event Ticket ASA...');
    
    // Get suggested parameters
    const suggestedParams = await algodClient.getTransactionParams().do();
    console.log('✅ Got suggested parameters');

    // Create ASA creation transaction
    const assetCreateTxn = algosdk.makeAssetCreateTxnWithSuggestedParamsFromObject({
      from: ORGANIZER_WALLET_ADDRESS,
      suggestedParams,
      
      // ASA Configuration
      total: 1000,  // Total supply of tickets
      decimals: 0,  // No decimals for tickets
      defaultFrozen: false,
      
      // Asset Names
      assetName: "AuthenTIX Launch Event Ticket",
      unitName: "TICKET",
      
      // Asset URLs and Metadata
      assetURL: "https://authentix.app/events/EVT1",
      assetMetadataHash: undefined,
      
      // Asset Control Addresses (all set to organizer for full control)
      manager: ORGANIZER_WALLET_ADDRESS,
      reserve: ORGANIZER_WALLET_ADDRESS,
      freeze: ORGANIZER_WALLET_ADDRESS,
      clawback: ORGANIZER_WALLET_ADDRESS,  // Important for anti-resell
      
      // Note
      note: new TextEncoder().encode("AuthenTIX Event Ticket - TestNet")
    });

    console.log('✅ ASA creation transaction prepared');
    console.log('\n📋 Transaction Details:');
    console.log('- Asset Name: AuthenTIX Launch Event Ticket');
    console.log('- Unit Name: TICKET');
    console.log('- Total Supply: 1000');
    console.log('- Decimals: 0');
    console.log('- Manager: ' + ORGANIZER_WALLET_ADDRESS);
    console.log('- Clawback: ' + ORGANIZER_WALLET_ADDRESS + ' (Anti-resell protection)');
    
    // Convert to base64 for signing
    const txnBase64 = Buffer.from(assetCreateTxn.toByte()).toString('base64');
    
    console.log('\n🔐 Transaction ready for signing:');
    console.log('Transaction (Base64):', txnBase64);
    
    console.log('\n📝 Next Steps:');
    console.log('1. Copy the transaction base64 above');
    console.log('2. Sign it with your Pera Wallet (organizer wallet)');
    console.log('3. Submit the signed transaction to get the ASA ID');
    console.log('4. Update your .env.local with: EVENT_ASA_ID=<your_asa_id>');
    console.log('5. Update the database with the ASA ID');
    
    return {
      txnBase64,
      assetCreateTxn
    };
    
  } catch (error) {
    console.error('❌ Error creating ASA:', error);
    throw error;
  }
}

// Run the script
createASA()
  .then((result) => {
    console.log('\n✅ ASA creation transaction prepared successfully!');
  })
  .catch((error) => {
    console.error('❌ Failed to create ASA:', error);
    process.exit(1);
  });
